rust/connlib/snownet/src/allocation.rs

```diff
@@ -1366,6 +1372,49 @@ mod tests {
         assert_eq!(channel, ChannelBindings::FIRST_CHANNEL);
     }
 
+    #[test]
+    fn uses_unused_channels_first_before_reusing_expired_one() {
+        let mut channel_bindings = ChannelBindings::default();
+        let mut now = Instant::now();
+
+        for _ in 0..100 {
+            let allocated_channel = channel_bindings.new_channel_to_peer(PEER1, now).unwrap();
+            channel_bindings.set_confirmed(allocated_channel, now);
+        }
+
+        now += Duration::from_secs(60 * 20); // All channels are expired and could be re-bound.
+
+        let channel = channel_bindings.new_channel_to_peer(PEER1, now).unwrap();
+
+        assert_eq!(channel, ChannelBindings::FIRST_CHANNEL + 100)
+    }
+
+    #[test]
+    fn uses_next_channel_as_long_as_its_available_before_reusing() {
+        let mut channel_bindings = ChannelBindings::default();
+        let mut now = Instant::now();
+
+        for _ in 0..=4095 {
+            let allocated_channel = channel_bindings.new_channel_to_peer(PEER1, now).unwrap();
+            channel_bindings.set_confirmed(allocated_channel, now);
+        }
+
+        now += Duration::from_secs(15 * 60); // All channels are expired and could be re-bound.
+        channel_bindings.set_confirmed(ChannelBindings::LAST_CHANNEL, now); // Last channel is in use
+
+        let channel = channel_bindings.new_channel_to_peer(PEER1, now).unwrap();
+        channel_bindings.set_confirmed(channel, now);
+
+        assert_eq!(channel, ChannelBindings::FIRST_CHANNEL);
+
+        now += Duration::from_secs(15 * 60); // All channels are expired and could be re-bound.
+        channel_bindings.set_confirmed(ChannelBindings::LAST_CHANNEL, now); // Last channel is in use
+        let channel = channel_bindings.new_channel_to_peer(PEER1, now).unwrap();
+
+        // We don't reuse the first channel, instead we should prefer the second channel
+        assert_ne!(channel, ChannelBindings::FIRST_CHANNEL)
```

Comment:
This is what I mean, since when we reach the maximum we always return first the first channel, we have a similar thing where we have a risk of a clock drift.

So basically, when the `LAST_CHANNEL` is occupied, our algorithm is equivalent to the one we had before